x-pack/plugins/apm/server/routes/agent_explorer/get_agents_filters.ts

```diff
@@ -0,0 +1,96 @@
+import { AGENT_NAME, AGENT_VERSION, SERVICE_LANGUAGE_NAME, SERVICE_NAME, SERVICE_NODE_NAME } from "@kbn/apm-plugin/common/elasticsearch_fieldnames";
+import { environmentQuery } from "@kbn/apm-plugin/common/utils/environment_query";
+import { ProcessorEvent } from "@kbn/observability-plugin/common";
+import { rangeQuery } from "@kbn/observability-plugin/server";
+import { RandomSampler } from "../../lib/helpers/get_random_sampler";
+import { Setup } from "../../lib/helpers/setup_request";
+
+const MAX_NUMBER_OF_SERVICES = 500;
+
+export async function getAgentsFilters({
+  environment,
+  setup,
+  start,
+  end,
+  randomSampler,
+}: {
+  environment: string;
+  setup: Setup;
+  start: number;
+  end: number;
+  randomSampler: RandomSampler;
+}) {
+const { apmEventClient } = setup;
+
+const response = await apmEventClient.search(
+  'get_agents_services',
+  {
+    apm: {
+      events: [ProcessorEvent.metric],
+    },
+    body: {
+      track_total_hits: false,
+      size: 0,
+      query: {
+        bool: {
+          filter: [
+            {
+              exists: {
+                field: AGENT_NAME,
+              }
+            },
+            {
+              exists: {
+                field: AGENT_VERSION,
+              }
+            },
+            {
+              exists: {
+                field: SERVICE_NODE_NAME,
+              }
+            },
+            ...rangeQuery(start, end),
+            ...environmentQuery(environment),
+          ],
+        },
+      },
+      aggs: {
+        sample: {
+          random_sampler: randomSampler,
+          aggs: {
+            services: {
+              terms: {
+                field: SERVICE_NAME,
+                size: MAX_NUMBER_OF_SERVICES,
+              },
+              aggs: {
+                sample: {
+                  top_metrics: {
+                    metrics: [
+                      { field: SERVICE_LANGUAGE_NAME } as const
+                    ],
+                    sort: {
+                      '@timestamp': 'desc' as const,
+                    }
+                  }
+                }
+              }
+            },
+          },
+        },
+      },
+    },
+  }
+);
+
+return (
+  response.aggregations?.sample.services.buckets.reduce((acc, service) => ({
```

Comment:
AFAIK terms agg returns unique values so I don't think it's necessary to create a `Set`

Reply:
In any case I will use the generic `suggestions`